core/src/main/java/com/datastax/oss/driver/internal/core/protocol/ByteBufCompressor.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (C) 2017-2017 DataStax Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.protocol;
+
+import com.datastax.oss.protocol.internal.Compressor;
+import io.netty.buffer.ByteBuf;
+import java.nio.ByteBuffer;
+
+public abstract class ByteBufCompressor implements Compressor<ByteBuf> {
+
+  @Override
+  public ByteBuf compress(ByteBuf uncompressed) {
+    return uncompressed.isDirect() ? compressDirect(uncompressed) : compressHeap(uncompressed);
```

Comment:
Nice to break up the direct/heap implementations, that made the code a bit hard to follow on 3.x. 👍